src/rust/engine/options/src/args.rs

```diff
@@ -3,95 +3,140 @@
 
 use std::env;
 
-use super::id::{NameTransform, OptionId, Scope};
+use super::id::{is_valid_scope_name, NameTransform, OptionId, Scope};
 use super::{DictEdit, OptionsSource};
-use crate::parse::{expand, expand_to_dict, expand_to_list, Parseable};
+use crate::parse::{expand, expand_to_dict, expand_to_list, ParseError, Parseable};
 use crate::ListEdit;
-use std::collections::HashMap;
-
-pub struct Args {
-    pub(crate) args: Vec<String>,
+use core::iter::once;
+use itertools::{chain, Itertools};
+
+#[derive(Debug)]
+struct Arg {
+    context: Scope,
+    flag: String,
+    value: Option<String>,
 }
 
-#[derive(PartialEq)]
-enum Negate {
-    True,
-    False,
-}
+impl Arg {
+    /// Checks if this arg's flag is equal to the provided strings concatenated with dashes.
+    /// E.g., "--foo-bar" matches ["-", "foo", "bar"].
+    fn flag_match<'a>(&self, dash_separated_strs: impl Iterator<Item = &'a str>) -> bool {
```

Comment:
Note that this does no allocation. The previous implementation generated temporary strings when checking if an option was referenced in the args.